Add the possibility to use a system wide installation of tl-optional (#207)
artivis/manif: CMakeLists.txt

#  INTERFACE_COMPILE_FEATURES cxx_std_11
#)

> option(USE_SYSTEM_WISE_TL_OPTIONAL "Use a system wise installation of tl-optional" OFF)

Reviewer: Is it 'system wide' or 'system wise' ? Not sure myself.

Author: I think It should be systemwide. 
https://en.wiktionary.org/wiki/systemwide

I  fix the typo

> if(USE_SYSTEM_WISE_TL_OPTIONAL)
>   find_package(tl-optional REQUIRED)
> endif()

Reviewer: A system installation of `tl-optional` is *only* used when this option is set. I'd rather prefer that it is used whenever available. Something along the lines of,
```suggestion
find_package(tl-optional 1.0.0 QUIET)
if(USE_SYSTEM_WISE_TL_OPTIONAL AND NOT tl-optional_FOUND)
  message(FATAL_ERROR "Could not find required package tl-optional")
endif()
```

target_include_directories(${PROJECT_NAME} INTERFACE

>   "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include$<$<NOT:$<BOOL:${USE_SYSTEM_WISE_TL_OPTIONAL}>>:;${PROJECT_SOURCE_DIR}/external/tl>>"

Reviewer: As per my previous comment, use `tl-optional_FOUND` instead of `USE_SYSTEM_WISE_TL_OPTIONAL`

Author: If `tl-optional_FOUND` is true and https://github.com/artivis/manif/pull/207#discussion_r649952287 is applied a systemwide installation of tl-optional will be used even if `USE_SYSTEM_WIDE_TL_OPTIONAL` is off. 

Is this one the desired behavior?

Reviewer: If the system tl-optional is at least v.1, then I'd say yes this is the desirable behavior.

  "$<INSTALL_INTERFACE:include>"
)

> target_link_libraries(${PROJECT_NAME} INTERFACE "$<$<BOOL:${USE_SYSTEM_WISE_TL_OPTIONAL}>:tl::optional>")

Reviewer: `USE_SYSTEM_WISE_TL_OPTIONAL`  -> `tl-optional_FOUND`

#  INTERFACE_COMPILE_FEATURES cxx_std_11
#)

# Check if there exists a system wise installation of tl-optional
find_package(tl-optional QUIET)

> # if tl-option exists manif will use the system wise installation otherwise it will provide it's own version
> if(tl-optional_FOUND)
> 
>   target_include_directories(${PROJECT_NAME} INTERFACE
>     "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>"
>     "$<INSTALL_INTERFACE:include>"
>     )
> 
>   target_link_libraries(${PROJECT_NAME} INTERFACE tl::optional)
> 
> else()
> 
>   target_include_directories(${PROJECT_NAME} INTERFACE
>     "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include;${PROJECT_SOURCE_DIR}/external/tl>"
>     "$<INSTALL_INTERFACE:include>"
>     )
> 
> endif()

Reviewer: The following is a little more concise. Not tested with an actual system-installed tl-optional,

```suggestion
target_include_directories(${PROJECT_NAME} INTERFACE
  "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include;$<$<NOT:$<BOOL:tl-optional_FOUND>>:${PROJECT_SOURCE_DIR}/external/tl>>"
  "$<INSTALL_INTERFACE:include>"
)

# if tl-option exists manif will use the system wise installation otherwise it will provide it's own version
if(tl-optional_FOUND)
  target_link_libraries(${PROJECT_NAME} INTERFACE tl::optional)
endif()
```

)

# Headers:

# if tl-optional is not found manif will install its own version

> if(NOT USE_SYSTEM_WISE_TL_OPTIONAL)

Reviewer: `USE_SYSTEM_WISE_TL_OPTIONAL`  -> `tl-optional_FOUND`

)

# Headers:

# if tl-optional is not found manif will install its own version
if(NOT tl-optional_FOUND)
  install(
    FILES "${PROJECT_SOURCE_DIR}/external/tl/tl/optional.hpp"
    DESTINATION include/tl

>     )

Reviewer: ```suggestion
  )
```
